lib/Sema/CompilerRepresentable.cpp

```diff
@@ -0,0 +1,325 @@
+                UnrepresentableType, UnderlyingReason->UnrepresentableType);
+    break;
+  }
+
+  assert(UnderlyingReason);
+  UnderlyingReason->emitDiagnosticNotesImpl(TC, Target);
+}
+
+void UnrepresentableReason::emitDiagnosticNotes(TypeChecker &TC,
+                                                const Decl *Target) const {
+  emitDiagnosticNotesImpl(TC, Target);
+}
+
+void UnrepresentableReason::emitDiagnosticNotes(TypeChecker &TC,
+                                                SourceLoc Target) const {
+  emitDiagnosticNotesImpl(TC, Target);
+}
+
+// Checks if the compile-time evaluator can handle the types that the generic
+// arguments are bound to.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkGenericArgs(CanBoundGenericType type) {
+  assert(type);
+  for (auto arg : type->getGenericArgs()) {
+    if (auto unrepresentable = check(arg)) {
+      return UnrepresentableReason::getGenericArgUnrepresentable(
+          type, std::move(unrepresentable));
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle this enum.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkDeclaredType(const EnumDecl *decl) {
+  assert(decl);
+  for (auto *elementDecl : decl->getAllElements()) {
+    auto *parameterList = elementDecl->getParameterList();
+    if (!parameterList)
+      continue;
+
+    for (auto *paramDecl : *parameterList) {
+      if (auto unrepresentable = check(paramDecl->getType())) {
+        return UnrepresentableReason::getEnumElementUnrepresentable(
+            decl->getDeclaredType(), elementDecl, std::move(unrepresentable));
+      }
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle this struct.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkDeclaredType(const StructDecl *decl) {
+  assert(decl);
+  for (auto *fieldDecl : decl->getStoredProperties()) {
+    if (auto unrepresentable = check(fieldDecl->getType())) {
+      return UnrepresentableReason::getFieldUnrepresentable(
+          decl->getDeclaredType(), fieldDecl, std::move(unrepresentable));
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle the types of the parameters
+// of this function.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::checkParams(CanAnyFunctionType type) {
+  assert(type);
+  for (auto param : type->getParams()) {
+    if (auto unrepresentable = check(param.getType())) {
+      return UnrepresentableReason::getParamUnrepresentable(
+          type, std::move(unrepresentable));
+    }
+  }
+  return NULL;
+}
+
+// Checks if the compile-time evaluator can handle values of type `type`.
+std::unique_ptr<const UnrepresentableReason>
+CompilerRepresentableChecker::check(Type type) {
+  if (!type)
+    return NULL;
+
+  auto canType = type->getCanonicalType();
+  if (isa<BuiltinIntegerType>(canType)) {
```

Comment:
Would it be a bit cleaner to use a `switch` on `TypeKind` here?

Reply:
There are lots of `dyn_cast`s, which are less clean to replicate in a `switch`. Also, there's one `dyn_cast` to a superclass (`AnyFunctionType`), which can't be done at all in a `switch`.

Reply:
I meant using `cast<>` directly in a switch case that requires the use of the cast value. I think it's cleaner because there's no long if-else chains. For example:
```cpp
switch (canType->getKind()) {
// allowed builtins
case BuiltinInteger:
// allowed generic types
case Archetype:
case GenericTypeParam:
  return nullptr;
case Struct:
  auto *decl = cast<StructType>(canType)->getDecl();
  ...
}
```

Both styles exist in the code base so I don't think there's a preference convention-wise.  And I don't have a strong opinion here.